subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+	}
+}
+
+/* allocation entry close (this closes the current sector) by writing offset
+ * of last ate to the sector end.
+ */
+static int zms_sector_close(struct zms_fs *fs)
+{
+	int rc;
+	struct zms_ate close_ate, garbage_ate;
+	size_t ate_size;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	close_ate.id = ZMS_HEAD_ID;
+	close_ate.len = 0U;
+	close_ate.offset = (uint32_t)((fs->ate_wra + ate_size) & ADDR_OFFS_MASK);
+	close_ate.metadata = 0xffffffff;
+
+	/* When we close the sector, we must write all non used ATE with
+	 * a non valid (Junk) ATE.
+	 * This is needed to avoid some corner cases where some ATEs are
+	 * not overwritten and become valid when the cycle counter wrap again
+	 * to the same cycle counter of the old ATE.
+	 * Example :
+	 * - An ATE.cycl_cnt == 0 is written as last ATE of the sector
+	   - This ATE was never overwritten in the next 255 cycles because of
+	     large data size
+	   - Next 256th cycle the leading cycle_cnt is 0, this ATE becomes
+	     valid even if it is not the case.
+	 */
+	close_ate.cycle_cnt = fs->sector_cycle;
+	memset(&garbage_ate, fs->flash_parameters->erase_value, sizeof(garbage_ate));
+	while (fs->ate_wra >= (fs->data_wra + ate_size)) {
+		rc = zms_flash_ate_wrt(fs, &garbage_ate);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	fs->ate_wra &= ADDR_SECT_MASK;
+	fs->ate_wra += (fs->sector_size - ate_size);
+
+	zms_ate_crc8_update(&close_ate);
+
+	(void)zms_flash_ate_wrt(fs, &close_ate);
+
+	/* skip the empty ATE */
+	fs->ate_wra -= ate_size;
+
+	zms_sector_advance(fs, &fs->ate_wra);
+
+	rc = zms_get_sector_cycle(fs, fs->ate_wra, &fs->sector_cycle);
+	if (rc == -ENOENT) {
+		/* sector never used */
+		fs->sector_cycle = 0;
+	} else if (rc) {
+		/* bad flash read */
+		return rc;
+	}
+
+	fs->data_wra = fs->ate_wra & ADDR_SECT_MASK;
+
+	return 0;
+}
+
+static int zms_add_gc_done_ate(struct zms_fs *fs)
+{
+	struct zms_ate gc_done_ate;
+
+	LOG_DBG("Adding gc done ate at %llx", fs->ate_wra);
+	gc_done_ate.id = ZMS_HEAD_ID;
+	gc_done_ate.len = 0U;
+	gc_done_ate.offset = (uint32_t)(fs->data_wra & ADDR_OFFS_MASK);
+	gc_done_ate.metadata = 0xffffffff;
+	gc_done_ate.cycle_cnt = fs->sector_cycle;
+
+	zms_ate_crc8_update(&gc_done_ate);
+
+	return zms_flash_ate_wrt(fs, &gc_done_ate);
+}
+
+static int zms_add_empty_ate(struct zms_fs *fs, uint64_t addr)
+{
+	struct zms_ate empty_ate;
+	size_t ate_size;
+	uint8_t cycle_cnt;
+	int rc = 0;
+
+	addr &= ADDR_SECT_MASK;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	LOG_DBG("Adding empty ate at %llx", (uint64_t)(addr + fs->sector_size - 2 * ate_size));
+	empty_ate.id = ZMS_HEAD_ID;
+	empty_ate.len = 0xffff;
+	empty_ate.offset = 0U;
+	empty_ate.metadata = ZMS_DEFAULT_VERSION;
+
+	rc = zms_get_sector_cycle(fs, addr, &cycle_cnt);
+	if (rc == -ENOENT) {
+		/* sector never used */
+		cycle_cnt = 0;
+	} else if (rc) {
+		/* bad flash read */
+		return rc;
+	}
+
+	/* increase cycle counter */
+	empty_ate.cycle_cnt = (cycle_cnt + 1) % BIT(8);
```

Comment:
Yeah, I am referencing this line here https://github.com/zephyrproject-rtos/zephyr/pull/77930/files#r1768412478